### ISSUE TITLE
no-set-state error does not trigger when in an arrow function's scope

### ISSUE THREAD

**Issue by kevinSuttle** [yannickcr/eslint-plugin-react#931]
Snippets where it is _not_ detected:

``` jsx
toggle = () => {
    const { visible } = this.state;
    // No error thrown
    this.setState({ visible: !visible });

    const hookEvent = !visible ? document.addEventListener : document.removeEventListener;
    hookEvent('mousedown', this.close);
    hookEvent('keydown', this.toggle);
  }
```

``` jsx
onMouseEnter={() => this.setState({ dropdownIndex: index })}
```

Snippet where it is detected and throws an error (same file as above)

``` jsx
keyDown = (e) => {
    const keycode = e.keyCode || e.Which,
      { dropdownIndex } = this.state

    switch (keycode) {
      case 13 /* ENTER */:
        if (this.state.optionsShown) {
          this.selectedItem(dropdownIndex)
        }
        break
      case 38 /* UP */:
      case 40 /* DOWN */:
        if (!this.state.optionsShown) {
          this.toggle()
        } else {
          const { options } = this.props,
            down = keycode === 40,
            adjustIndex = (index) => index < 0 ? options.length - 1 :
              index >= options.length ? 0 : index

         // Throws error, not in 
          this.setState({ dropdownIndex: adjustIndex(dropdownIndex + (down ? 1 : -1)) },
            this.focusCurrentItem)
        }
        return e.preventDefault()
    }
  }
```

**.eslintrc.json**

``` json
{
  "parser": "babel-eslint",
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module",
    "ecmaFeatures": {
      "classes": false,
      "jsx": true
    }
  },
  "plugins": [
    "import",
    "react",
    "jsx-a11y"
  ],
  "extends": [
    "airbnb",
    "eslint:recommended",
    "plugin:react/recommended"
  ],
  "globals": {
    "document": true,
    "window": true
  },
  "rules": {
    "react/no-unused-prop-types": ["warn", {"skipShapeProps": true}],
    "react/no-set-state": 2,
    "react/prefer-stateless-function": 2,
    "import/extensions": [2, "never", { "js": "never", "jsx": "never", "json": "never"}],
    "import/newline-after-import": 0,
    "import/no-extraneous-dependencies": [
      "off",
      {
        "devDependencies": false,
        "optionalDependencies": false
      }
    ],
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }]
  }
}
```

**Comment by kevinSuttle** [comment]
Thank you!
